libs/native-types/src/type_parameter.rs

```diff
@@ -0,0 +1,163 @@
+            s.parse().map(TypeParameter::Number).map_err(|_| {
+                let kind = io::ErrorKind::InvalidInput;
+                io::Error::new(kind, "Allowed inputs: `u64` or `max`")
+            })
+        }
+    }
+}
+
+impl Serialize for TypeParameter {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Number(num) => serializer.serialize_u64(*num),
+            Self::Max => serializer.serialize_str("Max"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for TypeParameter {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_any(TypeParameterVisitor)
+    }
+}
+
+struct TypeParameterVisitor;
+
+impl<'de> Visitor<'de> for TypeParameterVisitor {
+    type Value = TypeParameter;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "either `max` or a number")
+    }
+
+    fn visit_u8<E>(self, value: u8) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(TypeParameter::Number(value as u64))
+    }
+
+    fn visit_u16<E>(self, value: u16) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(TypeParameter::Number(value as u64))
+    }
+
+    fn visit_u32<E>(self, value: u32) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(TypeParameter::Number(value as u64))
+    }
+
+    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(TypeParameter::Number(value))
+    }
+
+    fn visit_i8<E>(self, value: i8) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        u64::try_from(value)
+            .map(TypeParameter::Number)
+            .map_err(|_| E::invalid_value(Unexpected::Signed(value as i64), &self))
+    }
+
+    fn visit_i16<E>(self, value: i16) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        u64::try_from(value)
+            .map(TypeParameter::Number)
+            .map_err(|_| E::invalid_value(Unexpected::Signed(value as i64), &self))
+    }
+
+    fn visit_i32<E>(self, value: i32) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        u64::try_from(value)
+            .map(TypeParameter::Number)
+            .map_err(|_| E::invalid_value(Unexpected::Signed(value as i64), &self))
+    }
+
+    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        u64::try_from(value)
+            .map(TypeParameter::Number)
+            .map_err(|_| E::invalid_value(Unexpected::Signed(value), &self))
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        value
+            .parse()
+            .map_err(|_| E::invalid_value(Unexpected::Str(value), &self))
+    }
+}
```

Comment:
maybe we could get away with `#[derive(Serialize, Deserialize)]` because we don't care about the shape, only that it can roundtrip